Get rid of Babel and rewrite @wdio/repl

christian-bromann opened the pull request:
## Proposed changes

closes #5795 
closes #5424

I took a stab at rewriting the smallest module to see how easy it is. Some things that need to get figured out:

- for some reasons unit test fail because the import [here](https://github.com/webdriverio/webdriverio/blob/master/packages/wdio-cli/src/commands/config.js#L9-L12) resolves in undefined for all
- to access private properties in tests I had to use `repl['_somePrivateIdentifier']` .. not sure if there is a better way to do this
- in our unit tests TypeScript things we use `expect-webdriverio` rather than the actual `expect` framework

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] I have added proper type definitions for new commands (if appropriate)

## Further comments

n/a

### Reviewers: @webdriverio/project-committers

christian-bromann commented on packages/wdio-repl/tests/repl.test.ts:
```diff
@@ -1,10 +1,15 @@
 import vm from 'vm'
 import replMock from 'repl'
+import expect from 'expect'
```
Any idea how I can get rid of this? Without this TypeScript believes we use `expect-webdriverio`. The problem importing `expect` is that it would be redundant and Jest would mock dependencies away that we have defined in this project.

erwinheitzman replied:
I feel like we should exclude the type in that case since apparently it is installed I think?

christian-bromann replied:
How would you exclude it?

erwinheitzman commented on tsconfig.json:
```diff
@@ -1,18 +1,19 @@
 {
   "compilerOptions": {
-    "target": "es5",
+    "target": "es2017",
```
This would cause us not able to run browsers like IE11 anymore, is this really what we want?

erwinheitzman replied:
If you want to use newer JavaScript then you need to define a lib instead of changing the output (target)

christian-bromann replied:
The WebdriverIO code only runs on Node.js. We have some scripts in the webdriverio package that we inject into the browser but they should not be compiled.

erwinheitzman commented on packages/webdriverio/tests/utils/interception/devtools.test.js:
```diff
@@ -78,7 +78,7 @@ test('abort request', async () => {
     }
     const mock = new NetworkInterception('**/foobar/**')
-    expect(::mock.abort).toThrow()
+    expect(mock.abort.bind(mock)).toThrow()
```
I feel like these binds are overkill, especially in a test. Is it not?

christian-bromann replied:
They are necessary otherwise we would loose the scope and the function would not work.

mgrybyk reviewed (changes requested):
Let's avoid using ts ignore if possible

christian-bromann commented on packages/wdio-repl/src/index.ts:
```diff
@@ -73,71 +81,79 @@ export default class WDIORepl {
-        return new Promise(
-            (resolve) => this.replServer.on('exit', resolve))
+        return new Promise((resolve) => {
+            if (!this._replServer) {
```
yeah, it can't be defined at initialisation because the interface only starts it when the start method is being executed.

christian-bromann replied:
Usually this is not necessary but TypeScript was not happy without checking

mgrybyk replied:
This is because replServer can be undefined according to line 53.
Anyway, it's also possible to use 'as'
(this._replServer as ReplServerTypeHere).on(...

christian-bromann commented:
@mgrybyk @erwinheitzman I addressed all issues and tests should pass now

> * for some reasons unit test fail because the import [here](https://github.com/webdriverio/webdriverio/blob/master/packages/wdio-cli/src/commands/config.js?rgh-link-date=2020-09-07T21%3A53%3A59Z#L9-L12) resolves in undefined for all

This was because of a circular dependency. Good to see that TS catches this.

> * in our unit tests TypeScript things we use `expect-webdriverio` rather than the actual `expect` framework

This problem resolved by itself.

Overall building the project and watching all files has become a bit slower and more CPU intensive. I would like to explore what we can do to improve this.

Please review @webdriverio/technical-steering-committee , if this lands we can release the first TS compiled WebdriverIO version.

christian-bromann commented:
Some further improvements:

- added a `tsconfig.prod.json` that compiles without source maps
- modified build script: instead of running 38 TS compiler processes it now runs only one with the parameter `-p @wdio/cli @wdio/utils ...`
- moved build and watch step into one file
- removed `.npmrc` file from all packages as we don't use it
- added tsconfig to .npmignore

@mgrybyk mind taking another review?

christian-bromann commented:
💯 please review @webdriverio/technical-steering-committee

mgrybyk commented on tsconfig.prod.json:
```diff
@@ -0,0 +1,23 @@
+{
+  "compilerOptions": {
```
Shall sourceMap be set to false here?

christian-bromann replied:
no need, it is false by default

christian-bromann replied:
but happy to add it to make it more clear

erwinheitzman commented on .eslintignore:
```diff
@@ -2,6 +2,5 @@
 /packages/**/node_modules
 /packages/node_modules
```
I think that `/packages/**/node_modules` already covers `/packages/node_modules`

christian-bromann replied:
I assume I added it because it did not. 🤷

erwinheitzman commented on .eslintrc.js:
```diff
@@ -1,44 +1,44 @@
+        'require-atomic-updates': 0
+    },
+    overrides: [{
+        files: ['*.ts', '*.tsx'],
```
we will never use tsx files in this project, will we?

christian-bromann replied:
I can remove the `tsx` entry.

erwinheitzman commented on .gitignore:
```diff
@@ -66,8 +66,8 @@ tests/typings/**/dist
 devtools.d.ts
-!babel.config.js
 !.eslintrc.js
+!jest.config.js
```
Why is this done?

christian-bromann replied:
I moved the Jest configurations from the package.json into a `jest.config.js`, given that we ignore all JS files in the root I had to ignore this new config file.

erwinheitzman commented on CONTRIBUTING.md:
````diff
@@ -77,13 +77,13 @@ If you only work on a single package you can watch only for that one by calling:
 
 ```sh
-$ npx jest ./packages/<package-name>/tests --watch
+$ npx jest ./packages/<package-name>/tests --watch --coverageReporters lcov
````
define the coverageReporters in the config file instead?

christian-bromann replied:
We do. This just changes the coverage reporter so you don't get to see this coverage list anymore and instead see the test results.

erwinheitzman commented on scripts/build.js:
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env node
```
why is this done over using the package.json files? for new people this adds complexity over using a simple setup that just uses lerna and the package.json files.

christian-bromann replied:
Running 1 TS compiler is much less CPU effective than 38. I tested both variants and this way it just works more efficient. At the end of the day it is the same command for the developer.

erwinheitzman replied:
Nice to know, thanks!
